src/mesh/duplicate_nodes.cc

```diff
@@ -0,0 +1,256 @@
+ * 2) Divide node_elements into groups (components) which share one of node_edges.
+ * 3) If more than one component was created, create for each extra component a new node
+ *    and update its index in the elements from this component.
+ *    Also set node_dim_ (dimension of elements using the node).
+ */
+void DuplicateNodes::duplicate_nodes()
+{
+  FOR_NODES(mesh_, n){
+    // create list of elements
+    std::vector<unsigned int> node_elements = mesh_->node_elements[n.index()];
+    
+    // create list of edges sharing the node
+    std::set<unsigned int> node_edges;
+    for (unsigned int el_idx : node_elements)
+    {
+      FOR_ELEMENT_SIDES(&mesh_->element[el_idx], sid)
+        node_edges.insert(mesh_->element[el_idx].side(sid)->edge_idx());
+    }
+    
+    // divide node_elements into components connected by node_edges
+    std::vector<std::set<unsigned int> > components;
+    while (node_elements.size() > 0)
+    {
+      // create component containing the first element in node_elements
+      std::queue<unsigned int> q;
+      q.push(node_elements[0]);
+      node_elements.erase(node_elements.begin());
+      std::set<unsigned int> component;
+      while (q.size() > 0) {
+        auto elem = mesh_->element.full_iter(&mesh_->element[q.front()]);
+        component.insert(elem.index());
+        // add to queue adjacent elements sharing one of node_edges
+        FOR_ELEMENT_SIDES(elem, sid) {
+          auto side = elem->side(sid);
+          if (node_edges.find(side->edge_idx()) != node_edges.end())
+          {
+            unsigned int esid;
+            FOR_EDGE_SIDES(side->edge(), esid) {
+              auto adj_el_idx = side->edge()->side(esid)->element().index();
+              if (adj_el_idx != elem.index())
+              {
+                auto it = std::find(node_elements.begin(), node_elements.end(), adj_el_idx);
+                if (it != node_elements.end()) {
+                  node_elements.erase(it);
+                  q.push(adj_el_idx);
+                }
+              }
+            }
+          }
+        }
+        q.pop();
+      }
+      components.push_back(component);
+    }
+    
+    // for each component except the first one
+    // create a duplicate of the node
+    if (components.size() > 0) {
+      node_dim_[obj_4_node_[n.index()]] = mesh_->element[*components[0].begin()].dim();
+      components.erase(components.begin());
+    }
+    for (auto component : components) {
+      unsigned int new_nid = nodes_.size();
+      nodes_.push_back(new_nid);
+      node_dim_.push_back(mesh_->element[*component.begin()].dim());
+      for (auto el_idx : component) {
+        switch (mesh_->element[el_idx].dim()) {
+          case 1:
+            {
+              MeshObject<1> *elem = &lines_[obj_4_el_[el_idx]];
+              for (unsigned int i=0; i<2; i++)
+                if (elem->nodes[i] == n.index())
+                  elem->nodes[i] = new_nid;
+              for (unsigned int fi=0; fi<2; fi++)
```

Comment:
Currently it is enough since currently, MeshObjects are only elements, their 1-codimensional faces and nodes. In the future, when we will have complete graph of n-faces, it would not be sufficient.